src/adapter/src/catalog/migrate.rs

```diff
@@ -155,13 +156,482 @@ pub(crate) async fn migrate(
+                            )),
+                        })
+                }
+                _ => unreachable!("subsource items must correlate to subsources"),
+            }
+
+            let mut subsource_item = source_map.remove(&export_id).expect("item exists");
+            subsource_item.create_sql = subsource_stmt.to_ast_string_stable();
+
+            let present = updated_items.insert(export_id, subsource_item);
+            assert_eq!(present, None, "each export only updated a single time");
+
+            if export_id < source.id() {
+                tracing::info!(
+                    "subsource {} has a GlobalId less than its primary source ({})",
+                    export_id,
+                    source.id()
+                );
+                needs_new_id.push(export_id);
+                source_ids_of_updated_subsources.insert(source.id());
+            }
+        }
+
+        // Update source definition to no longer include list of referenced
+        // subsources.
+        let primary_source_id = source.id();
+        let mut primary_source_item = source_map
+            .remove(&primary_source_id)
+            .expect("source exists");
+        let mut primary_source_stmt =
+            mz_sql_parser::parser::parse_statements(&primary_source_item.create_sql)
+                .expect("parsing persisted create_sql must succeed")
+                .into_element()
+                .ast;
+
+        match &mut primary_source_stmt {
+            Statement::CreateSource(create_source_stmt) => {
+                create_source_stmt.referenced_subsources = None;
+            }
+            _ => unreachable!("subsource items must correlate to subsources"),
+        }
+
+        primary_source_item.create_sql = primary_source_stmt.to_ast_string_stable();
+        let present = updated_items.insert(primary_source_id, primary_source_item);
+        assert_eq!(present, None, "each source only updated a single time");
+    }
+
+    tx.update_items(updated_items)?;
+
+    let mut remaining_updates = VecDeque::from_iter(needs_new_id.drain(..));
+
+    // If this item's ID must be moved forward, so too must every item that
+    // depends on it except for its primary source ID.
+    while let Some(id) = remaining_updates.pop_front() {
+        needs_new_id.push(id);
+        remaining_updates.extend(
+            conn_catalog
+                .state()
+                .get_entry(&id)
+                .used_by()
+                .iter()
+                .filter(|id| !source_ids_of_updated_subsources.contains(id))
+                .cloned(),
+        );
+    }
+
+    // Ensure that each ID is present only once and that it is in the greatest
+    // position.
+    let mut id_dedup = BTreeSet::new();
+    needs_new_id = needs_new_id
+        .into_iter()
+        .rev()
+        .filter(|id| id_dedup.insert(*id))
+        .collect();
+
+    assign_new_user_global_ids(tx, conn_catalog, now, needs_new_id)
```

Comment:
> all of an item's dependencies have `GlobalId`s less than its own.

This is wrong--I should have typed "dependents have `GlobalId`s less than its own" or "dependencies have `GlobalIds greater than its own."

I regret the typo!

Reply:
> This is wrong--I should have typed "dependents have GlobalIds less than its own" or "dependencies have `GlobalIds greater than its own."

I think this is backwards too though.

In `src/repr/src/global_id.rs` the docstring says 
> (items with greater `GlobalId`s can depend on those lesser but never the other way around)
which makes sense to me, since we'll create an object first with a lower id, then later another object will be created with a higher id that can depend on the first object.

Reply:
I'm also a little concerned that we don't have a test that covers this or proves the global-id-reordering is successful -- is that something we think the upgrade-tests should cover or do we need to add an explicit check?

Reply:
hah, ok––the original comment was right:

> all of an item's dependents have `GlobalId`s greater than its own.

This line from your comment confuses me:

> and then it looks like we've made needs_new_id ordered such that the IDs at the front of the vec need to be converted to IDs greater than those after them.

The items at the front of the vec are those that should have the smallest `GlobalId`s (they are the dependencies); those at the back should have greater `GlobalId`s (they are the dependents).

> I'm also a little concerned that we don't have a test that covers this or proves the global-id-reordering is successful -- is that something we think the upgrade-tests should cover or do we need to add an explicit check?

The upgrade tests cover this. Subsources would panic if their primary source didn't exist when they got created. There are upgrade tests that create PG sources and then upgrade them:

```
bin/mzcompose --dev --find platform-checks run default --scenario=UpgradeEntireMz --check=PgCdc 
```

Sadly this isn't working for me locally atm ([Slack](https://materializeinc.slack.com/archives/C01LKF361MZ/p1713972699539909)) but it is running in CI.

Reply:
Issue with running that commands was fortunately just a local issue.

Reply:
Gotcha, my comment came from what I think is happening with the `needs_new_id` vec above:

First we drain the vec into a double-ended queue, such that the queue starts with the ids of the subsources that currently have a `GlobalId` less than their primary source.

Then we iterate over this queue from the front, pushing each value we encounter into the `needs_new_id` vec. We add the dependents of each subsource to the end of the queue because we want these dependents to have higher global ids than the things they depend on (the subsource, to start).

So we end up with a `needs_new_id` vec in an order such that each subsequent element needs to have an id greater than the one before it.
Then we reverse this `needs_new_id` vec and filter out duplicate ids, such that the elements at the end must have ids less than those before them.

But this is opposite to the docstring on the `assign_new_user_global_ids` method, which says that the first element will have the smallest ID and the last element will have the greatest.

Let's sync live but I just wanted to write this out to make sure I could clarify

Reply:
I found the issue that was letting my upgrade tests pass with this bug in place. To resolve the issue, I've added asserts that all of an item's dependents have `GlobalId`s greater than itself, as well as adjusting the `PgCdc` platform test to be harder to fool in the upgrade scenario (added a new view with complex dependencies).